### PR TITLE
Tabular regression task and example

### LINKED ISSUE #905: Regression metrics raise error in semanticsegmentation task
#### What is your question?
I am trying to train a `SemanticSegmentation()` task. Some metrics (not the loss) such as the `torchmetrics.MeanSquaredError()` and other regression metrics don't work in the flash training loop. The following error is raised:
`RuntimeError: Only Tensors created explicitly by the user (graph leaves) support the deepcopy protocol at the moment`.

Is there a way to avoid this error?

#### Full Error Trace:
```
  File "/home/patrick/agro/weedetect/flash_training.py", line 43, in <module>
    trainer.finetune(model, datamodule=datamodule, strategy="freeze")
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/flash/core/trainer.py", line 165, in finetune
    return super().fit(model, train_dataloader, val_dataloaders, datamodule)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 552, in fit
    self._run(model)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 922, in _run
    self._dispatch()
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 990, in _dispatch
    self.accelerator.start_training(self)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 92, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 161, in start_training
    self._results = trainer.run_stage()
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1000, in run_stage
    return self._run_train()
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1049, in _run_train
    self.fit_loop.run()
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/base.py", line 111, in run
    self.advance(*args, **kwargs)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/fit_loop.py", line 200, in advance
    epoch_output = self.epoch_loop.run(train_dataloader)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/base.py", line 111, in run
    self.advance(*args, **kwargs)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 130, in advance
    batch_output = self.batch_loop.run(batch, self.iteration_count, self._dataloader_idx)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py", line 100, in run
    super().run(batch, batch_idx, dataloader_idx)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/base.py", line 111, in run
    self.advance(*args, **kwargs)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py", line 149, in advance
    self.batch_outputs[opt_idx].append(deepcopy(result.training_step_output))
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/copy.py", line 161, in deepcopy
    y = copier(memo)
  File "/home/patrick/miniconda3/envs/weedector/lib/python3.7/site-packages/torch/_tensor.py", line 55, in __deepcopy__
    raise RuntimeError("Only Tensors created explicitly by the user "
RuntimeError: Only Tensors created explicitly by the user (graph leaves) support the deepcopy protocol at the moment

```

#### What's your environment?

 - OS: Ubuntu 18
 - Packaging pip
 - Version 0.5.1

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #905 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/tabular/regression/model.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        lr_scheduler: LR_SCHEDULER_TYPE = None,
+        metrics: METRICS_TYPE = None,
+        learning_rate: float = 1e-2,
+        serializer: SERIALIZER_TYPE = None,
+        **tabnet_kwargs,
+    ):
+        self.save_hyperparameters()
+
+        cat_dims, cat_emb_dim = zip(*embedding_sizes) if embedding_sizes else ([], [])
+        model = TabNet(
```
Comment (ethanwharris) [inline flash/tabular/regression/model.py:65]: I guess there is some duplicated code here with the `TabularClassifier`, any way we can reduce that? Perhaps a `TabNet` adapter or something?

**Review by ethanwharris — APPROVED** [review]
Awesome, example looks great :smiley: Mind adding some docs / tests? (can be in a follow-up :smiley:)

File: flash/core/regression.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    ):
+        metrics = metrics or torchmetrics.MeanSquaredError()
+        loss_fn = loss_fn or F.mse_loss
+
+        return metrics, loss_fn
+
+    def to_metrics_format(self, x: torch.Tensor) -> torch.Tensor:
+        if getattr(self.hparams, "multi_label", False):
+            return torch.sigmoid(x)
+        return torch.softmax(x, dim=1)
```
Comment (Borda) [inline flash/core/regression.py:38]: heh this seems like a bug :D

File: flash/tabular/regression/model.py (hunk 1 above)
Comment (Borda) [reply flash/tabular/regression/model.py:65]: it is almost the same just the output is 1, which is hardcoded...

**Review by ananyahjha93 — APPROVED** [review]

**Pull request merged**